R/score_binary.R

```diff
@@ -15,18 +15,39 @@
 
 score_binary <- function(data,
                          forecast_unit,
-                         metrics) {
+                         metrics, ...) {
+
+  # get a list of the functions to be applied.
+  # this should be done by some helper function / more elegantly
+  metrics <- as.list(metrics)
   if ("brier_score" %in% metrics) {
-    data[, "brier_score" := brier_score(true_value, prediction),
-      by = forecast_unit
-    ]
+    names(metrics)[metrics == "brier_score"] <- "brier_score"
+    metrics["brier_score"] <- list(brier_score)
   }
-
   if ("log_score" %in% metrics) {
-    data[, "log_score" := logs_binary(true_value, prediction),
-      by = forecast_unit
-    ]
+    names(metrics)[metrics == "log_score"] <- "log_score"
+    metrics["log_score"] <- list(logs_binary)
   }
+  metrics <- metrics[sapply(metrics, is.function)]
+
+  # need to check the functions used here
+  # i.e. check the function has the relevant arguments.
+
+
+  lapply(seq_along(metrics), function(function_to_call, ...) {
+    metric_name <- names(metrics[function_to_call])
+    fun <- metrics[[function_to_call]]
+
+    fun_accepts_ellipsis <- any(names(formals(fun)) == "...")
+
+    if (fun_accepts_ellipsis) {
+      data[, (metric_name) := fun(true_value, prediction, ...)]
+    } else {
+      data[, (metric_name) := fun(true_value, prediction)]
```

Comment:
does this approach mean that only the bria score can be by forecast_unit? Why is this handled differently? Is this likely to generalise well>